File: src/main/java/com/code/tama/triggerapi/helpers/world/BlockUtils.java
/* (C) TAMA Studios 2025 */
package com.code.tama.triggerapi.helpers.world;

import net.minecraft.core.BlockPos;
import net.minecraft.util.Mth;
import net.minecraft.world.level.ChunkPos;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.LightLayer;
import net.minecraft.world.level.block.Blocks;
import net.minecraft.world.level.block.CarpetBlock;
import net.minecraft.world.level.block.SlabBlock;
import net.minecraft.world.level.block.SnowLayerBlock;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.state.properties.SlabType;

import com.code.tama.triggerapi.NativeLoader;
import com.code.tama.triggerapi.universal.UniversalCommon;
import com.code.tama.triggerapi.universal.UniversalServerOnly;

@SuppressWarnings("unused")
public class BlockUtils {

	static {
		NativeLoader.load("tts_native");
	}

	// -- Native (Rust) — pure math, no MC objects ------------------------------

	private static native int getPackedLight(int blockLight, int skyLight);
	private static native String getRelativeBlockPos(int bx, int by, int bz, int ox, int oy, int oz);
	private static native String fromChunkAndLocal(int chunkX, int chunkZ, int localX, int localY, int localZ);
	private static native float getReverseHeightModifier(int blockType, int snowLayers);
	private static native float getDifferenceInHeight(int fromType, int fromLayers, int toType, int toLayers);
	private static native long packBlockPos(int x, int y, int z);
	private static native double blockDistance(int x1, int y1, int z1, int x2, int y2, int z2);
	private static native boolean isWithinRadius(int x1, int y1, int z1, int x2, int y2, int z2, double radius);
	private static native boolean sameChunk(int x1, int z1, int x2, int z2);

	// -- Public API — same signatures as before --------------------------------

	/** Breaks a block at the given position. */
	public static void breakBlock(Level world, BlockPos pos) {
		UniversalServerOnly.Level.breakBlock(world, pos);
	}

	/** Converts a chunk + local position into a world BlockPos. */
	public static BlockPos fromChunkAndLocal(ChunkPos chunkPos, BlockPos localPos) {
		String[] parts = fromChunkAndLocal(chunkPos.x, chunkPos.z, UniversalCommon.Pos.x(localPos),
				UniversalCommon.Pos.y(localPos), UniversalCommon.Pos.z(localPos)).split(",");
		return new BlockPos(Integer.parseInt(parts[0]), Integer.parseInt(parts[1]), Integer.parseInt(parts[2]));
	}

	/** Height difference between two block states. */
	public static float getDifferenceInHeight(BlockState from, BlockState to) {
		return getDifferenceInHeight(blockTypeId(from), snowLayers(from), blockTypeId(to), snowLayers(to));
	}

	/**
	 * Height fraction for a block state (air=1, full=1, slab=0.5, carpet=1/16,
	 * snow=layers/8).
	 */
	public static float getHeightModifier(BlockState state) {
		// Still useful to expose directly for callers who already have a BlockState
		if (state.getBlock().equals(Blocks.AIR))
			return 1f;
		if (state.getBlock() instanceof SlabBlock && state.getValue(SlabBlock.TYPE) == SlabType.BOTTOM)
			return 0.5f;
		if (state.getBlock() instanceof SnowLayerBlock)
			return state.getValue(SnowLayerBlock.LAYERS) * 0.125f;
		if (state.getBlock() instanceof CarpetBlock)
			return 0.0625f;
		return 1.0f;
	}

	/** Clamped block light level at pos. */
	public static int getLight(Level level, BlockPos pos) {
		return Mth.clamp(level.getBrightness(LightLayer.BLOCK, pos), 0, 15);
	}

	/** Packed block+sky light value for rendering. */
	public static int getPackedLight(Level level, BlockPos pos) {
		int block = level.getBrightness(LightLayer.BLOCK, pos);
		int sky = level.getBrightness(LightLayer.SKY, pos);
		return getPackedLight(block, sky); // native does the bit shift
	}

	/** Adds an offset position to a base position and returns the result. */
	public static BlockPos getRelativeBlockPos(BlockPos basePos, BlockPos offsetPos) {
		String[] parts = getRelativeBlockPos(basePos.getX(), basePos.getY(), basePos.getZ(), offsetPos.getX(),
				offsetPos.getY(), offsetPos.getZ()).split(",");
		return new BlockPos(Integer.parseInt(parts[0]), Integer.parseInt(parts[1]), Integer.parseInt(parts[2]));
	}

	public static float getReverseHeightModifier(BlockState state) {
		return getReverseHeightModifier(blockTypeId(state), snowLayers(state));
	}

	/** Returns true if the block at pos matches the expected block type. */
	public static boolean isBlock(Level world, BlockPos pos, BlockState expected) {
		return world.getBlockState(pos).is(expected.getBlock());
	}

	/**
	 * Places a block only if the position is currently empty. Returns true on
	 * success.
	 */
	public static boolean placeBlock(Level world, BlockPos pos, BlockState state) {
		if (world.isEmptyBlock(pos)) {
			world.setBlock(pos, state, 3);
			return true;
		}
		return false;
	}

	// -- Bonus convenience methods (native-backed) -----------------------------

	/** Euclidean distance between two BlockPos. */
	public static double distance(BlockPos a, BlockPos b) {
		return blockDistance(a.getX(), a.getY(), a.getZ(), b.getX(), b.getY(), b.getZ());
	}

	/** True if b is within radius of a. Avoids sqrt. */
	public static boolean isWithinRadius(BlockPos a, BlockPos b, double radius) {
		return isWithinRadius(a.getX(), a.getY(), a.getZ(), b.getX(), b.getY(), b.getZ(), radius);
	}

	/** True if both positions are in the same 16×16 chunk column. */
	public static boolean sameChunk(BlockPos a, BlockPos b) {
		return sameChunk(a.getX(), a.getZ(), b.getX(), b.getZ());
	}

	/**
	 * Packs a BlockPos into a long using Minecraft's own encoding. Faster
	 * HashMap/HashSet key than boxing a BlockPos object.
	 */
	public static long pack(BlockPos pos) {
		return packBlockPos(pos.getX(), pos.getY(), pos.getZ());
	}

	// -- Internal helpers ------------------------------------------------------

	/**
	 * Maps a BlockState to the numeric block-type id used by native height
	 * functions.
	 */
	private static int blockTypeId(BlockState state) {
		if (state.getBlock().equals(Blocks.AIR))
			return 4;
		if (state.getBlock() instanceof SlabBlock && state.getValue(SlabBlock.TYPE) == SlabType.BOTTOM)
			return 1;
		if (state.getBlock() instanceof SnowLayerBlock)
			return 2;
		if (state.getBlock() instanceof CarpetBlock)
			return 3;
		return 0;
	}

	private static int snowLayers(BlockState state) {
		if (state.getBlock() instanceof SnowLayerBlock)
			return state.getValue(SnowLayerBlock.LAYERS);
		return 1;
	}
}

File: src/native/src/block_utils.rs
use jni::JNIEnv;
use jni::objects::JClass;
use jni::sys::{jint, jfloat, jlong, jdouble, jboolean, jstring};
use jni::objects::JString;

// ── GetPackedLight ────────────────────────────────────────────────────────────
// Pass the raw block/sky brightness values from Java — we do the bit math.
// Java: native int getPackedLight(int blockLight, int skyLight);
#[no_mangle]
pub extern "system" fn Java_com_code_tama_triggerapi_helpers_world_BlockUtilsNative_getPackedLight(
    _env: JNIEnv, _class: JClass,
    block_light: jint, sky_light: jint,
) -> jint {
    let b = block_light.clamp(0, 15);
    let s = sky_light.clamp(0, 15);
    (s << 20) | (b << 4)
}

// ── GetRelativeBlockPos ───────────────────────────────────────────────────────
// Java: native String getRelativeBlockPos(int bx, int by, int bz, int ox, int oy, int oz);
// Returns "x,y,z"
#[no_mangle]
pub extern "system" fn Java_com_code_tama_triggerapi_helpers_world_BlockUtilsNative_getRelativeBlockPos(
    mut env: JNIEnv, _class: JClass,
    bx: jint, by: jint, bz: jint,
    ox: jint, oy: jint, oz: jint,
) -> jstring {
    env.new_string(format!("{},{},{}", bx + ox, by + oy, bz + oz))
       .expect("couldn't create string")
       .into_raw()
}

// ── FromChunkAndLocal ─────────────────────────────────────────────────────────
// Java: native String fromChunkAndLocal(int chunkX, int chunkZ, int localX, int localY, int localZ);
// Returns "x,y,z"
#[no_mangle]
pub extern "system" fn Java_com_code_tama_triggerapi_helpers_world_BlockUtilsNative_fromChunkAndLocal(
    mut env: JNIEnv, _class: JClass,
    chunk_x: jint, chunk_z: jint,
    local_x: jint, local_y: jint, local_z: jint,
) -> jstring {
    let world_x = (chunk_x << 4) + local_x;
    let world_z = (chunk_z << 4) + local_z;
    env.new_string(format!("{},{},{}", world_x, local_y, world_z))
       .expect("couldn't create string")
       .into_raw()
}

// ── GetReverseHeightModifier ──────────────────────────────────────────────────
// Just 1 - heightModifier. Kept here to mirror BlockUtils structure.
// blockType: 0=full, 1=slab_bottom, 2=snow(layers 1-8), 3=carpet, 4=air
// Java: native float getReverseHeightModifier(int blockType, int snowLayers);
#[no_mangle]
pub extern "system" fn Java_com_code_tama_triggerapi_helpers_world_BlockUtilsNative_getReverseHeightModifier(
    _env: JNIEnv, _class: JClass,
    block_type: jint, snow_layers: jint,
) -> jfloat {
    1.0 - height_modifier(block_type, snow_layers)
}

// ── GetDifferenceInHeight ─────────────────────────────────────────────────────
// Java: native float getDifferenceInHeight(int fromType, int fromLayers, int toType, int toLayers);
#[no_mangle]
pub extern "system" fn Java_com_code_tama_triggerapi_helpers_world_BlockUtilsNative_getDifferenceInHeight(
    _env: JNIEnv, _class: JClass,
    from_type: jint, from_layers: jint,
    to_type: jint, to_layers: jint,
) -> jfloat {
    height_modifier(to_type, to_layers) - height_modifier(from_type, from_layers)
}

// ── PackBlockPos ──────────────────────────────────────────────────────────────
// Same Minecraft BlockPos.asLong() encoding. Kept here alongside block logic.
// Java: native long packBlockPos(int x, int y, int z);
#[no_mangle]
pub extern "system" fn Java_com_code_tama_triggerapi_helpers_world_BlockUtilsNative_packBlockPos(
    _env: JNIEnv, _class: JClass,
    x: jint, y: jint, z: jint,
) -> jlong {
    let x = x as i64;
    let y = y as i64;
    let z = z as i64;
    ((x & 0x3FFFFFF) << 38) | ((z & 0x3FFFFFF) << 12) | (y & 0xFFF)
}

// ── Distance ──────────────────────────────────────────────────────────────────
// Java: native double blockDistance(int x1, int y1, int z1, int x2, int y2, int z2);
#[no_mangle]
pub extern "system" fn Java_com_code_tama_triggerapi_helpers_world_BlockUtilsNative_blockDistance(
    _env: JNIEnv, _class: JClass,
    x1: jint, y1: jint, z1: jint,
    x2: jint, y2: jint, z2: jint,
) -> jdouble {
    let dx = (x2 - x1) as f64;
    let dy = (y2 - y1) as f64;
    let dz = (z2 - z1) as f64;
    (dx*dx + dy*dy + dz*dz).sqrt()
}

// ── IsWithinRadius ────────────────────────────────────────────────────────────
// Java: native boolean isWithinRadius(int x1,int y1,int z1, int x2,int y2,int z2, double radius);
#[no_mangle]
pub extern "system" fn Java_com_code_tama_triggerapi_helpers_world_BlockUtilsNative_isWithinRadius(
    _env: JNIEnv, _class: JClass,
    x1: jint, y1: jint, z1: jint,
    x2: jint, y2: jint, z2: jint,
    radius: jdouble,
) -> jboolean {
    let dx = (x2 - x1) as f64;
    let dy = (y2 - y1) as f64;
    let dz = (z2 - z1) as f64;
    (dx*dx + dy*dy + dz*dz <= radius * radius) as jboolean
}

// ── SameChunk ─────────────────────────────────────────────────────────────────
// Java: native boolean sameChunk(int x1, int z1, int x2, int z2);
#[no_mangle]
pub extern "system" fn Java_com_code_tama_triggerapi_helpers_world_BlockUtilsNative_sameChunk(
    _env: JNIEnv, _class: JClass,
    x1: jint, z1: jint,
    x2: jint, z2: jint,
) -> jboolean {
    ((x1 >> 4) == (x2 >> 4) && (z1 >> 4) == (z2 >> 4)) as jboolean
}

// ── internal shared ───────────────────────────────────────────────────────────
fn height_modifier(block_type: i32, snow_layers: i32) -> f32 {
    match block_type {
        4 => 1.0,
        1 => 0.5,
        2 => snow_layers as f32 * 0.125,
        3 => 0.0625,
        _ => 1.0,
    }
}